src/mo_frontend/typing.ml

```diff
@@ -476,6 +476,28 @@ and check_typ' env typ : T.typ =
       (List.map (fun (field : typ_field) -> field.it.id) fields);
     let fs = List.map (check_typ_field env sort.it) fields in
     T.Obj (sort.it, List.sort T.compare_field fs)
+  | AndT (typ1, typ2) ->
+    let t1 = check_typ env typ1 in
+    let t2 = check_typ env typ2 in
+    let t = T.glb t1 t2 in
```

Comment:
Can't this lead to non-termination if t1 or t2 is sibling recursive type that isn't known to be contractive yet?

What does:
 "type A = A and/or A"
 produce?

Reply:
(Type elaboration used to be syntactic with some checks, now it ain't.)

Reply:
Good question. Intersection and union are not productive themselves, so this produces:
```
type error [M0157], block contains non-productive definition A
```
I think this works fine because we do that check before elaborating the types. I added a couple of simple tests.
> (Type elaboration used to be syntactic with some checks, now it ain't.)
True, is that a problem?

Reply:
Ah, okay, this example happens to work, because in the prepass a named type is simply represented as its abstract name and union/intersection shortcuts the reflexive case.

Here is a problematic case:
```
type T = U and V; type U = {}; type V = {};
```
This crashes with an assertion failure, because glb computation needs to expand the type id and runs into type `Pre` in the prepass. Hm...

Reply:
Bingo, thanks for persevering.

Reply:
Yeah, I thought about making them more primitive, but then I'd rather bite and make them proper intersection and union types.

Another option I considered is to check that the operands aren't Pre, and if they are, give an "cannot compute intersection on forward types" error, similar to what we do for type inference in similar cases.

Reply:
Okay, I implemented the latter. Extending our type checker's "incompleteness" that way seems okay given that we're already incomplete in similar ways. And it's conservative, so we can do smarter later if this ever turns out to be relevant in practice.